### PR TITLE
JWT-auth

### PR DESCRIPTION
**Pull request by pranavs18**
- Enabling JWT Auth for Host-API
-Moving Godeps to .Godeps
-Changing build script to support cross compilation for linux

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
+	"github.com/golang/glog"
 	"github.com/rancherio/host-api/app/common"
 	"github.com/rancherio/host-api/config"
-	jwt "github.com/dgrijalva/jwt-go"
-	"net/http"
 )
 
+var fileBytes []byte
+
 func Auth(rw http.ResponseWriter, req *http.Request) bool {
```
Comment (ibuildthecloud) [inline auth/auth.go]: All of this logging needs to be removed as this will all be logged on each request.  It's excessive.

Comment (pranavs18) [inline auth/auth.go]: Ok, but shouldn't we atleast log the errors ?

File: auth/auth.go [hunk 2]
```diff
@@ -1,21 +1,50 @@
...
+	getToken := req.URL.Query().Get("token")
 
-	token, err := jwt.ParseFromRequest(req, func(token *jwt.Token) (interface{}, error) {
-		return config.Config.Key, nil
+	if len(getToken) == 0 {
+		return false
+	}
+	glog.Infoln("Enabling Authencation for Host_API")
+
+	if len(fileBytes) == 0 {
```
Comment (ibuildthecloud) [inline auth/auth.go]: The logic of caching should be done on startup of host-api.  The logical place would be in the config.Parse().  Besides just reading the bytes you should parse the bytes to a `*rsa.PublicKey` from `crypto/rsa`.  The current approach requires you to parse the rsa on each request.

File: auth/auth.go [hunk 3]
```diff
@@ -25,7 +54,10 @@ func Auth(rw http.ResponseWriter, req *http.Request) bool {
 		return false
 	}
 
-	if token.Claims["hostUuid"] != config.Config.HostUuid {
+	if token.Claims["reportedUuid"] != nil && token.Claims["reportedUuid"] != config.Config.HostUuid {
```
Comment (ibuildthecloud) [inline auth/auth.go]: Don't check reportedUuid.  Refer to your cattle PR for more about that.

File: auth/auth.go [hunk 4]
```diff
@@ -25,7 +54,10 @@ func Auth(rw http.ResponseWriter, req *http.Request) bool {
...
+		return false
+	}
+	if token.Claims["hostUuid"] != nil && token.Claims["hostUuid"] != config.Config.HostUuid {
```
Comment (ibuildthecloud) [inline auth/auth.go]: Why was this changed from the old condition?

```go
if token.Claims["hostUuid"] != config.Config.HostUuid {
        return false
}
```

This new code, if the hostUuid is not passed, you don't validate it.  The hostUuid is required and essential for authorization.

File: main.go [hunk 5]
```diff
@@ -47,6 +47,7 @@ func httpInterceptor(router http.Handler) http.Handler {
 		startTime := time.Now()
 
 		if !auth.Auth(w, req) {
+			glog.Infoln("Host_API validation failed")
```
Comment (ibuildthecloud) [inline main.go]: Remove

File: scripts/build [hunk 6]
```diff
@@ -22,4 +22,4 @@ fi
 echo export GOPATH=$GOPATH
 
 mkdir -p bin
-go build -o bin/host-api
+GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -o bin/host-api
```
Comment (ibuildthecloud) [inline scripts/build]: This forces the build to Linux which means if you are running on a Mac you could never run this.  Please remove this change.

File: stats/conn.go [hunk 7]
```diff
@@ -1,9 +1,10 @@
 package stats
 
 import (
-	"github.com/gorilla/websocket"
 	"net/http"
 	"strings"
+
+	"github.com/gorilla/websocket"
```
Comment (ibuildthecloud) [inline stats/conn.go]: Please don't change things in files that aren't necessary.

File: auth/auth.go [hunk 8]
```diff
@@ -1,21 +1,50 @@
...
+		}
+		fileBytes = file
+	} else {
+		glog.Infoln("Reading cached content")
+	}
+	token, err := jwt.Parse(getToken, func(token *jwt.Token) (interface{}, error) {
+		return fileBytes, nil
 	})
 
+	if err == nil && token.Valid {
```
Comment (will-chan) [inline auth/auth.go]: I'm a bit confused to this log statement.  You log that the authentication is successful if there are no errors and the token is valid.  However, after you log this, you have additional checks on the hostUuid that could prove to be invalid and return a false.  The log message would then become misleading.

Comment (pranavs18) [inline auth/auth.go]: Yeah i'll improve the log message there. I wanted to imply that a valid token has been found, but authentication can still be unsuccessful at that point since we have an additional check on the hostUuid.

File: main.go (hunk 5 above)
Comment (pranavs18) [inline main.go]: Removed !

File: auth/auth.go (hunk 3 above)
Comment (pranavs18) [inline auth/auth.go]: This was because of the corresponding Cattle change. Removed this change to have a check only on HostUuid.

File: auth/auth.go (hunk 4 above)
Comment (pranavs18) [inline auth/auth.go]: Again, because of the corresponding cattle change where the key was being passed as "ReportedUuid". Changed it back to original condition.

File: scripts/build (hunk 6 above)
Comment (pranavs18) [inline scripts/build]: Changed back to the original script, but i'll like to put a note somewhere on the wiki so that people debugging host-api on a MAC should use the cross-compilation commands if the host being used a linux VM.

File: auth/auth.go (hunk 2 above)
Comment (pranavs18) [inline auth/auth.go]: Added the PublicRSA parsing mechanism and moved the caching logic to config.go. PR updated.

File: stats/conn.go (hunk 7 above)
Comment (pranavs18) [inline stats/conn.go]: When i save this file, GO auto-formats it to separate packages in GO from the external packages.

File: config/config.go [hunk 9]
```diff
@@ -19,6 +25,7 @@ type config struct {
 }
 
 var Config config
+var FileBytes interface{}
```
Comment (ibuildthecloud) [inline config/config.go]: Don't save at a package level.  Create Config.KeyBytes and save there.

**Comment by ibuildthecloud** [comment]
@pranavs18 Please cleanup all the unneeded changes in the PR (all of the file mode changes and conn.go)

File: config/config.go (hunk 9 above)
Comment (pranavs18) [inline config/config.go]: Moved FileBytes to the Config structure.

**Comment by pranavs18** [comment]
@ibuildthecloud  Removed all file mode changes and conn.go.

**Comment by ibuildthecloud** [comment]
Please name the field Config.KeyBytes

**Comment by pranavs18** [comment]
@ibuildthecloud  If you are referring to Config.FileBytes , I already changed that. That's where you mentioned to not declare it at the package level.

**Comment by pranavs18** [comment]
@ibuildthecloud  - Renamed it to Config.KeyBytes.
